### PR TITLE
[POAE7-2735] AggregationHashTable implementation of uint32 and uint64

### PR DESCRIPTION
**Pull request by Deegue**
### What changes were proposed in this pull request?

1. Add uint32 and uint64 type support for `AggregationHashTable`. 
int32 and int64 is not supported in this PR.
int32 and int64 will be supported in `transferToAggKey` of `AggregationHashTable`. 
2. Add related test cases.
3. Some improvement like dedup and naming.

### Why are the changes needed?
HashTable Implementation.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
UTs.

### CODE REVIEW

File: src/cider/common/Arena.h [hunk 1]
```diff
@@ -28,7 +28,7 @@
 
 #include <boost/noncopyable.hpp>
 
-namespace DB {
+namespace cider {
```
[the remark this answers is not in the document]
Comment (Deegue) [reply src/cider/common/Arena.h]: done

File: src/cider/common/base/StringRef.h [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+#if defined(__aarch64__) && defined(__ARM_FEATURE_CRC32)
+#include <arm_acle.h>
+#define CRC_INT __crc32cd
+#endif
+
+/**
+ * The std::string_view-like container to avoid creating strings to find substrings in the
+ * hash table.
+ */
+struct StringRef {
```
[the remark this answers is not in the document]
Comment (Deegue) [reply src/cider/common/base/StringRef.h:55]: Yes, it's for string and serialized key implementation, and commented in the PR description.

File: src/cider/tests/CiderNewHashTableTest.cpp [hunk 3]
```diff
@@ -192,6 +192,66 @@ TEST_F(CiderNewHashTableTest, hashTableTest) {
   CHECK_EQ(fht.contains(key), false);
 }
 
+TEST_F(CiderNewHashTableTest, aggUInt32Test) {
+  using AggregatedHashTableWithUInt32Key = HashMap<uint32_t, Block, HashCRC32<uint32_t>>;
```
[the remark this answers is not in the document]
Comment (Deegue) [reply src/cider/tests/CiderNewHashTableTest.cpp:196]: Using `uintxx` for keys aims to resolve keys as index of HashTable, while the precision of `int32` and 'uint32' are the same. 

That's what is going to be added in `transferToAggKey`.

File: src/cider/common/base/StringRef.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (jikunshang) [reply src/cider/common/base/StringRef.h:55]: if it is not related to int32/64 hash table, please move to another PR.
try to Break your work into small, single-purpose patches if possible.

File: src/cider/tests/CiderNewAggHashTableTest.cpp [hunk 4]
```diff
@@ -260,6 +261,236 @@ TEST_F(CiderNewAggHashTableTest, aggInt16Test) {
...
+  uint32_t init_value_len = 20;
+  int8_t* init_value_ptr = allocator->allocate(init_value_len);
+  reinterpret_cast<int64_t*>(init_value_ptr)[0] = (int64_t)0;
+  reinterpret_cast<int32_t*>(init_value_ptr + 8)[0] = (int32_t)0;
+  reinterpret_cast<int32_t*>(init_value_ptr + 12)[0] =
+      std::numeric_limits<int32_t>::max();
+  reinterpret_cast<int32_t*>(init_value_ptr + 16)[0] =
+      std::numeric_limits<int32_t>::min();
+
+  cider::hashtable::AggregationHashTable aggregator(keys, init_value_ptr, init_value_len);
```
Comment (jikunshang) [inline src/cider/tests/CiderNewAggHashTableTest.cpp:287]: this should not name as `aggregator`

File: src/cider/tests/CiderNewAggHashTableTest.cpp [hunk 5]
```diff
@@ -260,6 +261,236 @@ TEST_F(CiderNewAggHashTableTest, aggInt16Test) {
...
+
+  // Row0:
+  // Generate a key = 1
+  int8_t* key1_ptr = allocator->allocate(6);
+  reinterpret_cast<bool*>(key1_ptr)[0] = (bool)false;
+  reinterpret_cast<uint32_t*>(key1_ptr + 2)[0] = (uint32_t)1;
```
Comment (jikunshang) [inline src/cider/tests/CiderNewAggHashTableTest.cpp:293]: why + 2, and where `1` comes from? Please avoid use magic number.

File: src/cider/tests/CiderNewHashTableTest.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jikunshang) [reply src/cider/tests/CiderNewHashTableTest.cpp:196]: So, hot to use Int32HashTable? is it support now? if not, please note it. if support, please give example.

File: src/cider/tests/CiderNewAggHashTableTest.cpp [hunk 6]
```diff
@@ -260,6 +261,236 @@ TEST_F(CiderNewAggHashTableTest, aggInt16Test) {
...
+  // Use get api and return value address
+  int8_t* value1_ptr = aggregator.get(key1_ptr);
+
+  // Check init value
+  CHECK_EQ(reinterpret_cast<int64_t*>(value1_ptr)[0], 0);
```
Comment (jikunshang) [inline src/cider/tests/CiderNewAggHashTableTest.cpp:298]: `reinterpret_cast<int64_t*>(value1_ptr)[0]` -> `*reinterpret_cast<int64_t*>(value1_ptr)`
same for other

Comment (jikunshang) [reply src/cider/tests/CiderNewAggHashTableTest.cpp:298]: and avoid us 0,8,12,,16, create a arg like offset_vector.

**Comment by Deegue** [comment]
@jikunshang Thanks for review! All comments above are handled in the last commit. All int types will be transferred to unsigned int in `transferToAggKey` which will be implemented soon.

**Review by jikunshang — APPROVED** [review]

**Review by spevenhe — APPROVED** [review]

**Pull request merged**